### PR TITLE
KNOX-2619 - HA dispatch should failover regardless of noFallback conf…

### PR DESCRIPTION
**Pull request by pzampino**
…ig until session is established

## What changes were proposed in this pull request?

Modified ConfigurableHaDispatch behavior wrt to failover when sticky sessions are enabled, but fallback (i.e., failover) is disabled, and the initial request fails. With this change, the dispatch will failover until a session is established, regardless of the fallback configuration.

## How was this patch tested?

Added tests in org.apache.knox.gateway.ha.dispatch.DefaultHaDispatchTest (testFailoverStickyOnFallbackOff, testFailoverStickyOnFallbackOff_SessionEstablished).

mvn -Ppackage,release clean install

### CODE REVIEW

**Review by lmccay — CHANGES_REQUESTED** [review]

File: gateway-provider-ha/src/main/java/org/apache/knox/gateway/ha/dispatch/ConfigurableHADispatch.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ private void setKnoxHaCookie(HttpServletRequest inboundRequest,
 
   protected void failoverRequest(HttpUriRequest outboundRequest, HttpServletRequest inboundRequest, HttpServletResponse outboundResponse, HttpResponse inboundResponse, Exception exception) throws IOException {
     /* check for a case where no fallback is configured */
-    if(stickySessionsEnabled && noFallbackEnabled) {
+    if(stickySessionsEnabled && noFallbackEnabled && inboundRequest.getCookies() != null) {
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply gateway-provider-ha/src/main/java/org/apache/knox/gateway/ha/dispatch/ConfigurableHADispatch.java:227]: That is a good point, there could be a chance that browser/client might send other cookies during the first request(which i did not think about). Checking for `stickySessionCookieName` cookie would make this more robust.

[the remark this answers is not in the document]
Comment (pzampino) [reply gateway-provider-ha/src/main/java/org/apache/knox/gateway/ha/dispatch/ConfigurableHADispatch.java]: I've added a check for the knox session cookie as the additional condition.

**Review by smolnar82 — COMMENTED** [review]
LGTM +1

**Review by moresandeep — APPROVED** [review]
lgtm, thanks @pzampino!

**Pull request merged**